text/0135-template-class-object-binding.md

```diff
@@ -0,0 +1,106 @@
+
+Popular UI frameworks offer such capability out of the box:
+
+- _Vue_: Use the standard [`class`](https://vuejs.org/guide/essentials/class-and-style) HTML attribute as described in this proposal.
+- _Svelte_: Use a custom template directive called [`class:name`](https://svelte.dev/docs/element-directives#class-name), to conditionally apply classes.
+- _Angular_: Use a custom template directive called [`[ngClass]`](https://angular.io/api/common/NgClass), with a similar API as described in this proposal.
+
+It is also worth mentioning that React has no built-in support for conditional style application. However, most React applications use the [classnames](https://www.npmjs.com/package/classnames) NPM package to fill this gap (6.3M weekly downloads).
+
+The `classSet` utility method exposed from the Salesforce internal `lightning/utils` module offers similar capabilities as the public classnames NPM package. This utility function is used by +500 components in Salesforce's internal code base. As this LWC module isn't exposed on the platform, platform components have to re-implement such utility methods for managing complex styles.
+
+As a side note, modern UI frameworks also have built-in capabilities to manipulate element styles. This capability is out of the scope of this RFC and will be the subject of a follow-up proposal.
+
+## Detailed design
+
+This RFC proposes to change the semantics of the template `class` attribute. On top of accepting a string value, the `class` attribute would now accept objects (plain objects and arrays) representing the CSS classes applied to the elements. The LWC engine would then interpret the object to determine which class names should be applied to the element.
+
+The list of class names applied to a given element is defined as follows:
+
+1. If the value is a `string`, apply all the classes in the value.
+2. If the value is an `object`, iterate over the enumerable string properties own by the object (excluding symbols and properties on the proto-chain) and apply the classes with the keys associated with a truthy value.
+3. If the value is an `array`, iterate over its entries and apply this algorithm recursively for each of its entries.
+4. Ignore all other values: `null`, `undefined`, other primitive values, and complex objects.
+
+| Input                                       | Output          |
+| ------------------------------------------- | --------------- |
+| `'foo bar'`                                 | `'foo bar'`     |
+| `{foo: true, bar: false, 'fiz buz': true }` | `foo fiz buz`   |
+| `{truthy: 1, falsey: '', alsoTruthy': Symbol() }` | `truthy alsoTruthy`   |
+| `[false, 'bar', null]`                      | `'bar'`         |
+| `['foo', { bar: true }, ['baz']]`           | `'foo bar baz'` |
```

Comment:
How valuable is it to allow recursion? It does introduce additional complexity into the implementation.

Reply:
Question answered: `classnames` in React is recursive, plus if you are mixing arrays and objects as in the first example, it's nicer to be able to put constant values in the array and non-constant values in the object.